src/main/java/com/puppycrawl/tools/checkstyle/checks/TrailingCommentCheck.java

```diff
@@ -200,76 +188,100 @@ public boolean isCommentNodesRequired() {
-                comment = commentList.get(commentList.size() - 1);
-                lineBefore = line.substring(0, comment.getStartColNo());
+        if (line.length() > ast.getLastChild().getColumnNo() + 1) {
+            line = line.substring(ast.getLastChild().getColumnNo() + 2);
+        }
+        final Pattern removeSlash = Pattern.compile("/");
+        line = removeSlash.matcher(line).replaceAll("");
 
-                // do not check comment which doesn't end line
-                if (comment.getText().length == 1
-                        && !CommonUtil.isBlank(line
-                            .substring(comment.getEndColNo() + 1))) {
-                    continue;
-                }
-            }
-            if (!format.matcher(lineBefore).find()
-                && !isLegalComment(comment)) {
-                illegalLines.put(lineNo, comment.getStartColNo());
-            }
+        final String lineBefore = getLines()[lineNo - 1].substring(0, ast.getColumnNo());
+
+        // do not check comment which doesn't end line
+        if (!(ast.getLineNo() == ast.getLastChild().getLineNo()
+                && !CommonUtil.isBlank(line)) && !format.matcher(lineBefore).find()
+                && !isCorrectBlockComment(ast, comment)) {
+            log(ast, MSG_KEY);
         }
     }
 
     /**
-     * Checks if given comment is legal (single-line and matches to the
-     * pattern).
+     * Checks if block comment is legal and matches to the pattern.
      *
+     * @param ast Detail ast element to be checked.
      * @param comment comment to check.
      * @return true if the comment if legal.
      */
-    private boolean isLegalComment(final TextBlock comment) {
+    private boolean isCorrectBlockComment(DetailAST ast, final String comment) {
         final boolean legal;
 
         // multi-line comment can not be legal
-        if (legalComment == null || comment.getStartLineNo() != comment.getEndLineNo()) {
+        if (legalComment == null
+                || ast.getFirstChild().getLineNo() != ast.getLastChild().getLineNo()) {
             legal = false;
         }
         else {
-            String commentText = comment.getText()[0];
-            // remove chars which start comment
-            commentText = commentText.substring(2);
-            // if this is a C-style comment we need to remove its end
-            if (commentText.endsWith("*/")) {
-                commentText = commentText.substring(0, commentText.length() - 2);
-            }
+            String commentText = comment;
             commentText = commentText.trim();
             legal = legalComment.matcher(commentText).find();
         }
         return legal;
     }
 
+    /**
+     * Checks if given single line comment is legal (single-line and matches to the
+     * pattern).
+     *
+     * @param comment comment to check.
+     * @return true if the comment if legal.
+     */
+    private boolean checkSingleLine(String comment) {
```

Comment:
javadoc and method name are confusing because there is 
```
    /**
     * Check is single line comment is legal.
     *
     * @param ast Detail ast element to be checked.
     */
    private void checkSingleLineComment(DetailAST ast) {
```

Reply:
Done,  please comment if new one is good.